lib/src/solver/result.dart

```diff
@@ -105,6 +94,35 @@ class SolveResult {
   SolveResult(this._root, this._previousLockFile, this.packages, this.pubspecs,
       this.availableVersions, this.attemptedSolutions, this.resolutionTime);
 
+  /// Checks that the SolveResult is compatible with [_previousLockfile]
+  ///
+  /// Throws if pubspec.yaml isn't satisfied.
+  Future<void> enforceLockfile() async {
+    for (final package in packages) {
+      if (package.isRoot) continue;
+      final previousPackage = _previousLockFile.packages[package.name];
+      if (previousPackage == null) {
+        throw ApplicationException(
+            'Dependency ${package.name} is not already locked in `pubspec.lock`.');
+      }
+      if (previousPackage != package) {
+        throw ApplicationException(
+            'Dependency ${package.name} is locked to $previousPackage in `pubspec.lock` but resolves to $package.');
+      }
+      final previousDescription = previousPackage.description;
+      final newDescription = package.description;
+
+      if (previousDescription is ResolvedHostedDescription) {
+        final newHash = (newDescription as ResolvedHostedDescription).sha256;
+
+        if (!bytesEquals(previousDescription.sha256, newHash)) {
+          throw ApplicationException(
+              'Dependency ${package.name} is locked to a different content-hash than what was resolved.');
+        }
+      }
```

Comment:
Enforce lock file should not allow packages to be removed.
We should also have a test case for this.